### PR TITLE
Problem (Fix #1881): wallet sync logic is not compatible with the refactored tendermint-rs light client design

### LINKED ISSUE #1881: Problem: wallet sync logic is not compatible with the refactored tendermint-rs light client design
When thinking about #828 and trying to incorporate the new light client design of tendermint-rs, I think our wallet sync logic should be overhauled too.
- light client from tendermint-rs handled the all the light client sync and verification now, including all the disk and network io, running in standalone thread.
- it supports multiple peers, a big step on #1114.
- the light client sync process can be shared by different wallets.

wallet sync logic will be like this:
```rust
let light_client_handle = get_shared_light_client_handle_or_spawn();
let light_block = light_client_handle.verify_to_highest();
// download blocks and sync wallet states
// can safely do batch and fast forward here
// still need to verify block hash and app hashes and the chain
// and verify last block matches the light_block
// only persist wallet state after synced to the last block
set wallet.trusted to false
sync_untrusted_to(light_block.height, light_block.app_hash);
check hashes match with light_block
set wallet.trsuted to true
```

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- running light client superivsor, and sync wallet to recent
  trusted block

### CODE REVIEW

**Comment by leejw51crypto** [comment]
where is light client supervisor code ?
so , it's run as a separate process ?

**Comment by tomtau** [comment]
> where is light client supervisor code ?
> so , it's run as a separate process ?

it's in the Tendermint refactored light-client crate (refactored from "tendermint-rs" crate) -- see https://github.com/informalsystems/tendermint-rs/blob/master/docs/architecture/adr-006-light-client-refactor.md#integration

**Comment by tomtau** [comment]
https://github.com/informalsystems/tendermint-rs/blob/master/light-client/src/supervisor.rs#L113
https://github.com/informalsystems/tendermint-rs/blob/master/docs/architecture/adr-007-light-client-supervisor-ergonomics.md

**Comment by yihuang** [comment]
The ring build issue happen again:
```
[build:325] error[E0425]: cannot find function `fill_impl` in this scope
[build:326]    --> /drone/src/drone/cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.12/src/rand.rs:160:9
[build:327]     |
[build:328] 160 |         fill_impl(dest)
[build:329]     |         ^^^^^^^^^ help: try: `self.fill_impl`
[build:330]
[build:331]    Compiling regex v1.3.9
[build:332] error: aborting due to previous error
[build:333]
[build:334] For more information about this error, try `rustc --explain E0425`.
[build:335] error: could not compile `ring`.
[build:336]
[build:337] To learn more, run the command again with --verbose.
[build:338] warning: build failed, waiting for other jobs to finish...
[build:339] error: build failed
```

**Comment by devashishdxt** [comment]
This pull request downgrades `ring` back to `0.16.12`. Maybe update rebase with `master`.

**Comment by devashishdxt** [comment]
It also downgrades `webpki`.

**Comment by yihuang** [comment]
> This pull request downgrades `ring` back to `0.16.12`. Maybe update rebase with `master`.

The cause of the issue is because tendermint-rs bring in signatory-secp256k1, which has conflict dependencies with newer version of ring. 
https://github.com/informalsystems/tendermint-rs/issues/391#issuecomment-656575144

So either change tendermint-rs to make signatory-secp256k1 optional, or downgrade our ring for now.

**Comment by leejw51crypto** [comment]
Supervisor term is confusing with that of python , some comment about it?

**Comment by yihuang** [comment]
> Supervisor term is confusing with that of python , some comment about it?

added the tendermint-rs design doc in comment.

File: Cargo.lock [hunk 1]
```diff
@@ -797,7 +814,9 @@ dependencies = [
  "secstr",
  "serde",
  "serde_json",
+ "sled 0.31.0",
```
Comment (tomtau) [inline Cargo.lock:817]: is it necessary to have both sled 0.31 and 0.33?

File: Cargo.toml [hunk 2]
```diff
@@ -55,3 +55,15 @@ default-members = [
 ring = { git = "https://github.com/crypto-com/ring.git", rev = "bdbcc7041095f028d49d9fecd7edcf26d6083274" }
 # FIXME: use upstream when merged
 sha2 = { git = "https://github.com/crypto-com/hashes.git", rev = "289d5b76f2163a3808010341ed1df3cb156d97e1" }
+
+[patch.crates-io.tendermint]
```
Comment (tomtau) [inline Cargo.toml:59]: any PRs waiting for upstream?

File: client-core/Cargo.toml [hunk 3]
```diff
@@ -34,12 +34,14 @@ tokio = { version = "0.1.22", default-features = false, features = ["rt-full"] }
...
-ring = "0.16.12"
-tendermint = { git = "https://github.com/crypto-com/tendermint-rs.git", default-features = false, rev = "ceca4219d6ae4e4ce906a6579e865af15458fdd6" }
+ring = "0.16.15"
+tendermint = "0.14"
+tendermint-light-client = "0.14"
 thiserror = { version = "1.0", default-features = false }
 non-empty-vec = "0.1"
 zxcvbn = "2.0"
 indexmap = "1.4"
+sled = "0.31.0"
```
Comment (tomtau) [inline client-core/Cargo.toml:44]:
```suggestion
sled = "0.33.0"
```

File: client-core/src/service/sync_state_service.rs [hunk 4]
```diff
@@ -14,20 +11,23 @@ pub struct SyncState {
     pub last_block_height: u64,
     /// last app hash
     pub last_app_hash: String,
-    /// current trusted state for lite client verification
-    pub trusted_state: lite::TrustedState,
+    /// last block hash
+    pub last_block_hash: String,
```
Comment (tomtau) [inline client-core/src/service/sync_state_service.rs:15]: `String` is more compatible with the light client APIs?
`SyncState` isn't directly returned from RPC, so these could have more precise types / no need to store the hex string?

File: client-core/src/wallet/syncer.rs [hunk 5]
```diff
@@ -750,6 +860,56 @@ fn filter_incomming_staking_transactions<'a>(
...
+
+    let timeout = Duration::from_secs(10);
+    let io = ProdIo::new(peer_map, Some(timeout));
+
+    let db = sled::open(db_path).unwrap_or_else(|e| {
+        println!("[ error ] could not open database: {}", e);
+        std::process::exit(1);
+    });
+
+    let mut light_store = SledStore::new(db);
```
Comment (tomtau) [inline client-core/src/wallet/syncer.rs:879]: so the light client has a separate storage? any guarantees that this won't conflict with client's internal storage (e.g. have some root path and always post-fix it?)?

File: client-core/Cargo.toml (hunk 3 above)
Comment (tomtau) [inline client-core/Cargo.toml:44]:
```suggestion
sled = "0.33.0"
```

File: Cargo.lock (hunk 1 above)
Comment (yihuang) [reply Cargo.lock:817]: tendermint-light-client still using sled 0.31 (https://github.com/informalsystems/tendermint-rs/blob/master/light-client/Cargo.toml#L21)

File: client-core/src/service/sync_state_service.rs (hunk 4 above)
Comment (yihuang) [reply client-core/src/service/sync_state_service.rs:15]: No, it's just `SyncState` already using `String` to represent hashes, so I kept that tradition...

File: client-core/src/wallet/syncer.rs (hunk 5 above)
Comment (yihuang) [reply client-core/src/wallet/syncer.rs:879]: There's separate db for each peer, the db path is `{wallet_db_path}/{peer_id}/`
https://github.com/crypto-com/chain/pull/1901/files#diff-873b16429c2cd004665e7ef924cf4643R1272

File: client-core/src/wallet/syncer.rs [hunk 6]
```diff
@@ -1101,3 +1256,81 @@ mod tests {
...
+    db_path: &Path,
+    addr: &str,
+) -> LightClientHandle<impl Handle + 'static> {
+    // FIXME handle it better
+    let addr = format!("tcp://127.0.0.1:{}", &addr[15..20]);
+    let addr = tendermint::net::Address::from_str(&addr).unwrap();
+
+    let primary: PeerId = "BADFADAD0BEFEEDC0C0ADEADBEEFC0FFEEFACADE".parse().unwrap();
+    let witness: PeerId = "CEFEEDBADFADAD0C0CEEFACADE0ADEADBEEFC0FF".parse().unwrap();
+
```
[the remark this answers is not in the document]
Comment (yihuang) [reply client-core/src/wallet/syncer.rs:1271]: It's supposed to be the tendermint node id, added a `FIXME` to improve it.
We can pass the node id from cli arguments, also witness node informations.

File: client-core/src/wallet/syncer.rs [hunk 7]
```diff
@@ -819,6 +981,7 @@ mod tests {
     }
 
     #[test]
+    #[ignore]
```
[the remark this answers is not in the document]
Comment (yihuang) [reply client-core/src/wallet/syncer.rs:984]: Forgot about it, enabling it now.

File: Cargo.toml (hunk 2 above)
Comment (yihuang) [reply Cargo.toml:59]: the pr is merged now, the patch is changed too.

File: client-rpc/src/handler.rs [hunk 8]
```diff
@@ -63,11 +65,27 @@ impl RpcHandler {
             fee_policy.clone(),
             tendermint_client.clone(),
         )?;
+        // wait for first block to appear
+        for _ in 0..10 {
```
Comment (tomtau) [inline client-rpc/src/handler.rs:69]: should it quit if after 10 attempts there aren't any blocks, or it'll fail anyway?

File: client-core/src/wallet/syncer.rs [hunk 9]
```diff
@@ -750,6 +865,56 @@ fn filter_incomming_staking_transactions<'a>(
...
+        verification_trace: HashMap::new(),
+    };
+
+    let options = light_client::Options {
+        trust_threshold: TrustThreshold {
+            numerator: 1,
+            denominator: 3,
+        },
+        trusting_period: Duration::from_secs(36000),
+        clock_drift: Duration::from_secs(1),
```
Comment (tomtau) [inline client-core/src/wallet/syncer.rs:907]: any comment on these options? I guess threshold of 1/3 is natural.
for the period -- should it be equal to the evidence time / unbonding time?

File: client-core/src/wallet/syncer.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomtau) [reply client-core/src/wallet/syncer.rs]: Will this fail if the PeerID is different, or is this only for convention / storage?

File: client-rpc/src/handler.rs (hunk 8 above)
Comment (yihuang) [reply client-rpc/src/handler.rs:69]: removed this waiting, user should just restart when bad things happen.

File: client-core/src/wallet/syncer.rs (hunk 9 above)
Comment (yihuang) [reply client-core/src/wallet/syncer.rs:907]: added some comments and links

**Comment by tomtau** [comment]
bors try

File: Cargo.lock (hunk 1 above)
Comment (yihuang) [reply Cargo.lock]: upstream is updated, no sled 0.31 anymore.

File: Cargo.toml [hunk 10]
```diff
@@ -55,3 +55,15 @@ default-members = [
...
+git = "https://github.com/informalsystems/tendermint-rs.git"
+branch = "master"
+
+[patch.crates-io.tendermint-rpc]
+git = "https://github.com/informalsystems/tendermint-rs.git"
+branch = "master"
+
+[patch.crates-io.tendermint-light-client]
+git = "https://github.com/informalsystems/tendermint-rs.git"
+branch = "master"
```
Comment (tomtau) [inline Cargo.toml:69]: it's now published on crates.io https://crates.io/crates/tendermint-light-node ?

Comment (tomtau) [reply Cargo.toml:69]: https://crates.io/crates/tendermint-light-client
https://crates.io/crates/tendermint
https://crates.io/crates/tendermint-rpc

File: client-core/src/wallet/syncer.rs [hunk 11]
```diff
@@ -188,20 +232,21 @@ fn load_view_key<S: SecureStorage>(storage: &S, name: &str, enckey: &SecKey) ->
...
 }
 
-impl<S, C, O, T> WalletSyncer<S, C, TxObfuscationDecryptor<O>, T>
+impl<S, C, O, T, L> WalletSyncer<S, C, TxObfuscationDecryptor<O>, T, L>
 where
     S: SecureStorage,
     C: Client,
     O: TransactionObfuscation,
     T: AddressRecovery,
+    L: Handle + Send + Sync + Clone,
```
Comment (tomtau) [inline client-core/src/wallet/syncer.rs:241]: This `Handle + Send + Sync + Clone` is quite repeated, so maybe give a alias:
```
type LightClient = Handle + Send + Sync + Clone;
```
?

Comment (yihuang) [reply client-core/src/wallet/syncer.rs]: done, made an extension trait for that.

File: Cargo.toml (hunk 10 above)
Comment (yihuang) [reply Cargo.toml]: done, removed patch.

**Comment by tomtau** [comment]
bors try

**Comment by yihuang** [comment]
> ## try
> Build failed:
> 
> * [drone-hw/push](http://13.82.172.109/crypto-com/chain/1014)

I don't see this issue running locally
```
Preparing... WITH_FEE
--
201 | Error: Run enclave error: No such file or directory (os error 2)
```

The error message come from `dev-utils keypackage generate`, it don't find enclave file, strange.

**Comment by tomtau** [comment]
> > ## try
> > Build failed:
> > 
> > * [drone-hw/push](http://13.82.172.109/crypto-com/chain/1014)
> 
> I don't see this issue running locally
> 
> ```
> Preparing... WITH_FEE
> --
> 201 | Error: Run enclave error: No such file or directory (os error 2)
> ```
> 
> The error message come from `dev-utils keypackage generate`, it don't find enclave file, strange.

not sure if it's because of the force-push while it was executing

**Comment by tomtau** [comment]
bors retry

**Comment by tomtau** [comment]
> > 
> > The error message come from `dev-utils keypackage generate`, it don't find enclave file, strange.
> 
> not sure if it's because of the force-push while it was executing

still -- not sure. it doesn't need to be the enclave file; it can also be the signature or the "ftxsgx-runner"

**Comment by yihuang** [comment]
> > > The error message come from `dev-utils keypackage generate`, it don't find enclave file, strange.
> > 
> > 
> > not sure if it's because of the force-push while it was executing
> 
> still -- not sure. it doesn't need to be the enclave file; it can also be the signature or the "ftxsgx-runner"

it should be the `PATH` env issue, and my local `chain-test` image is different from the one in the hub. fixed.

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
bors retry

**Comment by tomtau** [comment]
bors try-

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
bors try-

**Pull request merged**